WooCommerce/Classes/ViewRelated/Products/Filters/FilterProductListViewModel.swift

```diff
@@ -0,0 +1,126 @@
+import UIKit
+import Yosemite
+
+/// `FilterListViewModel` for filtering a list of products.
+final class FilterProductListViewModel: FilterListViewModel {
+    /// Aggregates the filter values that can be updated in the Filter Products UI.
+    struct Filters {
+        let stockStatus: ProductStockStatus?
+        let productStatus: ProductStatus?
+        let productType: ProductType?
+    }
+
+    let filterActionTitle = NSLocalizedString("Show Products", comment: "Button title for applying filters to a list of products.")
+
+    let filterTypeViewModels: [FilterTypeViewModel]
+
+    private let stockStatusFilterViewModel: FilterTypeViewModel
+    private let productStatusFilterViewModel: FilterTypeViewModel
+    private let productTypeFilterViewModel: FilterTypeViewModel
+
+    /// Used to dismiss the Filter Products UI.
+    private let sourceViewController: UIViewController
```

Comment:
⚠️  I think we shouldn't have this. The `ViewModel` should be data-only as much as possible. But in here, we're actually giving some control to the `FilterProductListViewModel` when to _dismiss_ something. One could even do something like this in this `ViewModel`:

```swift
😱 
sourceViewController.navigationController?.present(....) 
```

We don't want that. 😅 

My suggestion is to move this control to the `FilterListViewController` instead. First, we can define what the final value is:

```swift
protocol FilterListViewModel {
    /// The type of the final value returned to the caller of FilterListViewController
    associatedtype Criteria

    /// The final value returned to the caller of FilterListViewController
    var criteria: Criteria { get }
```

And implement this in `FilterProductListViewModel` like:

```swift
    typealias Criteria = Filters

    var criteria: Filters {
        let stockStatus = self.stockStatusFilterViewModel.selectedValue as? ProductStockStatus ?? nil
        let productStatus = self.productStatusFilterViewModel.selectedValue as? ProductStatus ?? nil
        let productType = self.productTypeFilterViewModel.selectedValue as? ProductType ?? nil

        return Filters(stockStatus: stockStatus, productStatus: productStatus, productType: productType)
    }

```

The `FitlerListVC` has to define this too:

```swift
final class FilterListViewController<ViewModel: FilterListViewModel>: UIViewController {
```

And then `FilterListVC` can have the `onCompletion` callback:

```swift
    /// Callback when the user tapped on the filter button. The argument will be `nil` if the user
    /// only pressed Dismiss.
    ///
    private let onCompletion: (ViewModel.Criteria?) -> ()
```

And then during the button presses, the VC would **retrieve** the data to return from the `ViewModel`:

```swift
    @objc func filterActionButtonTapped() {
        dismiss(animated: true) {
            let criteria = self.viewModel.criteria
            self.onCompletion(criteria)
        }
    }

    @objc func dismissButtonTapped() {
        dismiss(animated: true) {
            self.onCompletion(nil)
        }
    }
```

The call site, `ProductsViewController` would look like this:

```swift
        let viewModel = FilterProductListViewModel(filters: filters)
        let filterProductListViewController = FilterListViewController(viewModel: viewModel, onCompletion: { [weak self] criteria in
            if let criteria = criteria {
                self?.filters = criteria
            }
            // TODO-2037: filter products based on the filters
        })
        present(filterProductListViewController, animated: true, completion: nil)
```

This seems more in line with how UI messaging works in iOS. The `ViewControllers` **control** the navigation and dismissal. The `ViewModel` would only operate on the data. 🙂 

Here is a diff that you should be able to apply in this PR's branch to test this out: https://gist.github.com/shiki/226492bf67a6d328afa72a19373f8d82.

Reply:
I like your suggestion! updated in: https://github.com/woocommerce/woocommerce-ios/pull/2204/commits/bb8f875ac8957e98b0def8fe7c127c2de149a324